src/gsCore/gsBoundary.h

```diff
@@ -428,12 +407,196 @@ struct patchCorner : public boxCorner
+public:
+
+    /// @brief Constructor
+    ///
+    /// @param b          The index that defines the component
+    /// @param total_dim  The dimension of the box itself
+    boxComponent( index_t b, index_t total_dim ) : m_index(b), m_total_dim(total_dim) {}
+
+    /// @brief Constructor creating \a boxComponent representing the interior
+    ///
+    /// @param total_dim  The dimension of the box itself
+    boxComponent( index_t total_dim ) : m_index(0), m_total_dim(total_dim) {}
+
+    /// @brief Constructor converting \a boxSide to a boxComponent
+    ///
+    /// @param b          The box side
+    /// @param total_dim  The dimension of the box itself
+    boxComponent( boxSide b, index_t total_dim );
+
+    /// @brief Constructor converting \a boxCorner to a boxComponent
+    ///
+    /// @param b          The box corner
+    /// @param total_dim  The dimension of the box itself
+    boxComponent( boxCorner b, index_t total_dim );
+
+    /// Dimension of the computational domain (the box itself)
+    index_t totalDim() const { return m_total_dim; }
+
+    /// Dimension of the component
+    index_t dim() const;
+
+    /// Returns the index
+    index_t index() const { return m_index; }
+
+    /// Returns a vector of all \a boxCorner s that are contained in the component
+    std::vector<boxCorner> containedCorners() const;
+
+    /// Returns a vector of all \a boxSide s that contain the component
+    std::vector<boxSide> containingSides() const;
+
+    /// Converts to \a boxSide and fails if the component is not a side
+    boxSide asSide() const;
+
+    /// Converts to \a boxCorner and fails if the component is not a corner
+    boxCorner asCorner() const;
+
+    /// Returns the opposite boxCorner
+    boxComponent opposite() const;
+
+    /// Represents a location
+    enum location {
+        interior = 0,  ///< Represents the interior
+        begin = 1,     ///< Represents the beginning
+        end = 2        ///< Represents the end
+    };
+
+    /// Gets the location for the direction
+    ///
+    /// If the result value is \a begin, then the component is characterized
+    /// by \$f x_i = 0 \$f
+    /// If the result value is \a end, then the component is characterized
+    /// by \$f x_i = 1 \$f
+    /// If the result value is \a begin, then the component is characterized
+    /// by \$f x_i \in (0,1) \$f
+    ///
+    /// @param direction   The index \$f i \$f from above
+    location locationForDirection(index_t direction) const;
+
+    /// Sets the location for the direction
+    ///
+    /// See \a locationForDirection
+    ///
+    /// @param direction   The index \$f i \$f from above
+    void setLocationForDirection(index_t direction, location par);
+
+};
+
+/**
+ *   @brief Struct which represents a certain component (interior, face, egde, corner) of a particular patch
+ *
+ *   This struct is a generalization of \a patchSide and of \a patchCorner.
+ *
+ */
+
+struct GISMO_EXPORT patchComponent : boxComponent {
```

Comment:
Maybe adding m_patch as a member of boxComponent (with default value 0) and completely removing patchComponent simplifies things ? 
Or is there a real advantage to have them separately ?

(I pose it as a question but we can do this, if needed, in a new PR, so not to be changed in this PR)

Reply:
I have followed the design of box/patchCorner and box/patchSide.

I think an alternative design (like what you have proposed) is possible.